api/features/versioning/views.py

```diff
@@ -0,0 +1,162 @@
+                project__in=self.request.user.get_permitted_projects(VIEW_PROJECT)
+            ),
+            pk=self.kwargs["feature_pk"],
+        )
+        environment = get_object_or_404(
+            self.request.user.get_permitted_environments(
+                VIEW_ENVIRONMENT, project=feature.project
+            ),
+            pk=self.kwargs["environment_pk"],
+        )
+
+        request.feature = feature
+        request.environment = environment
+
+    def get_queryset(self):
+        if getattr(self, "swagger_fake_view", False):
+            return EnvironmentFeatureVersion.objects.none()
+
+        return EnvironmentFeatureVersion.objects.filter(
+            environment=self.request.environment, feature_id=self.request.feature
+        )
+
+    def perform_create(self, serializer: Serializer) -> None:
+        serializer.save(
+            environment=self.request.environment, feature=self.request.feature
+        )
+
+    def perform_destroy(self, instance: EnvironmentFeatureVersion) -> None:
+        if instance.is_live:
+            raise FeatureVersionDeleteError("Cannot delete a live version.")
+
+        super().perform_destroy(instance)
+
+    @action(detail=True, methods=["POST"])
+    def publish(self, request: Request, **kwargs) -> Response:
+        ef_version = self.get_object()
+        serializer = self.get_serializer(data=request.data, instance=ef_version)
+        serializer.is_valid(raise_exception=True)
+        serializer.save(published_by=request.user)
+        return Response(serializer.data)
+
+
+class EnvironmentFeatureVersionFeatureStatesViewSet(
+    GenericViewSet,
+    ListModelMixin,
+    CreateModelMixin,
+    UpdateModelMixin,
+    DestroyModelMixin,
+):
+    serializer_class = EnvironmentFeatureVersionFeatureStateSerializer
+    permission_classes = [
+        IsAuthenticated,
+        EnvironmentFeatureVersionFeatureStatePermissions,
+    ]
+    pagination_class = None
+
+    def get_queryset(self):
+        if getattr(self, "swagger_fake_view", False):
+            return FeatureState.objects.none()
+
+        environment_feature_version_sha = self.kwargs["environment_feature_version_pk"]
+
+        return FeatureState.objects.filter(
+            environment=self.request.environment,
+            feature=self.request.feature,
+            environment_feature_version_id=environment_feature_version_sha,
+        )
+
+    def initial(self, request: Request, *args, **kwargs):
+        super().initial(request, *args, **kwargs)
+        environment_feature_version = get_object_or_404(
+            EnvironmentFeatureVersion, sha=self.kwargs["environment_feature_version_pk"]
+        )
+        if self.action != "list" and environment_feature_version.published is True:
+            raise FeatureVersionDeleteError("Cannot modify published version.")
+
+        # patch the objects onto the request
+        request.environment_feature_version = environment_feature_version
```

Comment:
What are your thoughts on adding these objects to `self` instead of request if they are just being used by the view? Is patching request, a convention for sharing data between multiple methods in a view class? I think request having a large surface area that goes beyond the view makes this a bit hard to track?

Reply:
think we are adding them to the context using `  def get_serializer_context(self):`, no?

Reply:
Ah yes, this was why I changed it: 

<img width="755" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/d27c4b36-e5ea-40cb-853d-af9ca9b4acba">

I've moved them to the view itself but, to satisfy my IDE (and I guess python conventions), I have also overridden the `__init__` method to set them to `None`.

Reply:
Oh, interesting...